cclib/method/ddec.py

```diff
@@ -274,24 +274,56 @@ def calculate(self, indices=None, fupdate=0.05):
         self.logger.info("Conditioning charge densities.")
         self.condition_densities()
 
+        # STEP 4-7
+        # Optimize weights on each grid point
+        kappa = [0.0 for x in self.data.atomnos]
+        update_kappa = False
+        self.N_A = []
+        steps = 4
+        while steps < 5:  # change this to 8 after implementing kappa routine
+            self.logger.info("Optimizing charge weights (Step {}).".format(steps))
+            self.N_A.append(self._calculate_w_and_u())
+
+            # Update kappa as described in Figure S4.2 of doi: 10.1039/c6ra04656h
+            if numpy.any([x if x < -1e-5 else 0 for x in self.N_A[-1]]) and steps != 4:
+                update_kappa = True
+            if (
+                steps > 5
+                and update_kappa
+                and numpy.diff(self.N_A)[-1] < 1e-5
+                and numpy.diff(self.N_A)[-2] < 1e-5
+            ):
+                update_kappa = False
+                kappa = [0.0 for x in self.data.atomnos]
+
+            # TODO: EXIT LOOP HERE for steps = 7
```

Comment:
if you change the if statement below to `if update_kappa and steps < 7:` then it will just exit the loop if steps == 7

Reply:
I was trying to wait until the full implementation of steps 4-7 (with PR921) before I add the exit conditions here. The exit condition is present in PR921 now. Thanks!